pkg/distributor/distributor.go

```diff
@@ -415,12 +539,46 @@ func (d *Distributor) sendSamplesErr(ctx context.Context, ingester ring.Instance
 
 	_, err = c.(logproto.PusherClient).Push(ctx, req)
 	d.ingesterAppends.WithLabelValues(ingester.Addr).Inc()
-	if err != nil {
-		d.ingesterAppendFailures.WithLabelValues(ingester.Addr).Inc()
+
+	if err == nil {
+		return nil
+	}
+
+	d.ingesterAppendFailures.WithLabelValues(ingester.Addr).Inc()
+
+	if d.cfg.ShardStreams.Mode == NeverShardMode {
+		return err
+	}
+
+	// TODO: look for a better way to recognize a per-stream rate limit.
+	if errMsg := err.Error(); strings.Contains(errMsg, "Per stream rate limit exceeded (limit") {
+		d.streamSharding.Inc()
+		shardLimitedStream(errMsg, d.streamSharder)
 	}
 	return err
```

Comment:
Should there be a retry?

Reply:
agreed, good catch. I was taking a look at it and apparently it won't be too easy; I'll dig into it a little bit more today. Depending on the complexity I'd prefer doing it in a follow-up PR to make reviewing this one easier, but let's see!

Reply:
We removed the trigger that to split the query from this PR so it can be its own